streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -179,21 +180,22 @@ public void onCompletion(final RecordMetadata metadata,
                         offsets.put(tp, metadata.offset());
                     } else {
                         if (sendException == null) {
-                            if (exception instanceof ProducerFencedException) {
+                            if (exception instanceof ProducerFencedException || exception instanceof UnknownProducerIdException) {
                                 log.warn(LOG_MESSAGE, topic, exception.getMessage(), exception);
 
                                 // KAFKA-7510 put message key and value in TRACE level log so we don't leak data by default
                                 log.trace("Failed message: (key {} value {} timestamp {}) topic=[{}] partition=[{}]", key, value, timestamp, topic, partition);
 
-                                sendException = new ProducerFencedException(
+                                sendException = new RecoverableClientException(
                                     String.format(
                                         EXCEPTION_MESSAGE,
                                         logPrefix,
                                         "producer got fenced",
                                         timestamp,
                                         topic,
                                         exception.toString()
-                                    )
+                                    ),
+                                    exception
```

Comment:
From looking at the production log files, I think keeping the whole stack trace of the original `exception` is not necessary since we know exactly it would be thrown from

```
	at org.apache.kafka.streams.processor.internals.RecordCollectorImpl$1.onCompletion(RecordCollectorImpl.java:202)
	at org.apache.kafka.clients.producer.KafkaProducer$InterceptorCallback.onCompletion(KafkaProducer.java:1318)
	at org.apache.kafka.clients.producer.internals.ProducerBatch.completeFutureAndFireCallbacks(ProducerBatch.java:230)
	at org.apache.kafka.clients.producer.internals.ProducerBatch.done(ProducerBatch.java:196)
	at org.apache.kafka.clients.producer.internals.Sender.failBatch(Sender.java:730)
	at org.apache.kafka.clients.producer.internals.Sender.failBatch(Sender.java:716)
	at org.apache.kafka.clients.producer.internals.Sender.completeBatch(Sender.java:674)
	at org.apache.kafka.clients.producer.internals.Sender.handleProduceResponse(Sender.java:596)
	at org.apache.kafka.clients.producer.internals.Sender.access$100(Sender.java:74)
	at org.apache.kafka.clients.producer.internals.Sender$1.onComplete(Sender.java:798)
	at org.apache.kafka.clients.ClientResponse.onComplete(ClientResponse.java:109)
	at org.apache.kafka.clients.NetworkClient.completeResponses(NetworkClient.java:561)
	at org.apache.kafka.clients.NetworkClient.poll(NetworkClient.java:553)
	at org.apache.kafka.clients.producer.internals.Sender.runOnce(Sender.java:335)
	at org.apache.kafka.clients.producer.internals.Sender.run(Sender.java:244)
```

So may be we can just log the original exception name; similarly in `recordSendError` we can just keep the original exception's name but not the stack trace since we know exactly it is from the same `Sender.failBatch` trace.

Reply:
In practice, the exceptions we're recording here are ApiExceptions, so they don't contain stack traces anyway.

I actually added the exception to the "cause" on purpose, because it was a pain trying to understand the stack traces we were previously logging while I was digging in to this issue. Even if the log messages are more verbose, they're easy to follow. I.e., it's a nice chain of "E1 caused by E2 caused by E3", instead of "E1-with-E3-in-the-message caused by E2".

I actually wanted to take it farther by removing the `exception.toString` from the message, but I didn't want to be _too_ bold in this patch, and cause even more controversy.

Reply:
SG!